src/common/qat.c

```diff
@@ -0,0 +1,439 @@
+}
+
+static inline CpaPhysicalAddr
+virtToPhys(void *virtAddr)
+{
+	return (CpaPhysicalAddr)qaeVirtToPhysNUMA(virtAddr);
+}
+
+static void
+getDcInstance(CpaInstanceHandle *pDcInstHandle)
+{
+	CpaInstanceHandle dcInstHandles[MAX_INSTANCES];
+	Cpa16U numInstances = 0;
+	CpaStatus status = CPA_STATUS_SUCCESS;
+
+	*pDcInstHandle = NULL;
+	status = cpaDcGetNumInstances(&numInstances);
+
+	if (numInstances >= MAX_INSTANCES)
+		numInstances = MAX_INSTANCES;
+
+	if ((status == CPA_STATUS_SUCCESS) && (numInstances > 0)) {
+		status = cpaDcGetInstances(numInstances, dcInstHandles);
+		if (status == CPA_STATUS_SUCCESS)
+			*pDcInstHandle = dcInstHandles[0];
+	}
+}
+
+static void
+dcCallback(void *pCallbackTag, CpaStatus status)
+{
+	if (pCallbackTag != NULL)
+		*(Cpa32U *)pCallbackTag = 1;
+}
+
+/** Common Functions */
+int
+qat_dc_is_available()
+{
+	Cpa16U numInstances = 0;
+	CpaStatus status = icp_sal_userStartMultiProcess("SSL", CPA_FALSE);
+
+	if (status != CPA_STATUS_SUCCESS)
+		return 0;
+
+	cpaDcGetNumInstances(&numInstances);
+	icp_sal_userStop();
+
+	return (numInstances > 0);
+}
+
+/** Compression Functions */
+int
+qat_dc_init(CpaInstanceHandle *dcInstHandle,
+	    CpaDcSessionHandle *sessionHdl,
+	    Cpa16U *numInterBuffLists,
+	    CpaBufferList ***bufferInterArrayPtr,
+	    Cpa32U maxBufferSize,
+	    CpaDcCompLvl compLvl)
+{
+	CpaStatus status = CPA_STATUS_SUCCESS;
+	CpaDcInstanceCapabilities cap = {0};
+	CpaBufferList **interBufs = NULL;
+	Cpa16U i = 0;
+	Cpa32U buffMetaSize = 0;
+
+	Cpa32U sess_size = 0;
+	Cpa32U ctx_size = 0;
+	CpaDcSessionSetupData sd = {0};
+
+	*numInterBuffLists = 0;
+
+	status = qaeMemInit();
+	if (status != CPA_STATUS_SUCCESS) {
+		D_ERROR("QAT: Failed to initialize memory driver\n");
+		return -1;
+	}
+
+	status = icp_sal_userStartMultiProcess("SSL", CPA_FALSE);
+	if (status != CPA_STATUS_SUCCESS) {
+		D_ERROR("QAT: Failed to start user process SSL\n");
+		qaeMemDestroy();
+		return -1;
+	}
+
+	getDcInstance(dcInstHandle);
+	if (*dcInstHandle == NULL) {
+		D_ERROR("QAT: No DC instance\n");
+		return -1;
+	}
+
+	status = cpaDcQueryCapabilities(*dcInstHandle, &cap);
+	if (status != CPA_STATUS_SUCCESS)
+		return -1;
+
+	if (!cap.statelessDeflateCompression ||
+	    !cap.statelessDeflateDecompression ||
+	    !cap.checksumAdler32 ||
+	    !cap.dynamicHuffman)
+		return -1;
+
+	if (cap.dynamicHuffmanBufferReq) {
+		status = cpaDcBufferListGetMetaSize(*dcInstHandle, 1,
+						    &buffMetaSize);
+
+		if (status == CPA_STATUS_SUCCESS)
+			status = cpaDcGetNumIntermediateBuffers(
+				*dcInstHandle,
+				numInterBuffLists);
+
+		if (status == CPA_STATUS_SUCCESS && *numInterBuffLists != 0)
+			status = Mem_Alloc_Contig(
+				(void *)&interBufs,
+				*numInterBuffLists * sizeof(CpaBufferList *),
+				1);
+
+		/** if max buffer size is zero, set default value 64KB */
+		if (maxBufferSize == 0)
+			maxBufferSize = MAX_BUF_SIZE;
+
+		for (i = 0; i < *numInterBuffLists; i++) {
+			if (status == CPA_STATUS_SUCCESS)
```

Comment:
general style I've seen in daos is if status != SUCCESS then inline return.

Reply:
Need to clean up memories in the end of the function, so the check of status != SUCCESS can't return.